File: skills/seedream-image-generation/seedream.py
import os
import sys
import json
import urllib.request
import urllib.error
import argparse

def generate_image(
    prompt,
    model=None,
    api_key=None,
    size=None,
    watermark=None,
    optimize_prompt_options=None,
    tools=None,
    output_format=None,
    sequential_image_generation=None,
    download_dir=None
):
    """
    Generates an image using the Volcengine Seedream API.
    
    Args:
        prompt (str): The text description of the image to generate. Required.
        model (str): The target model endpoint ID (e.g., ep-xxxx). Required.
        api_key (str, optional): Custom API key to access Volcengine Ark. Defaults to None.
        size (str, optional): Target image size. Can be a resolution class (e.g., '2K', '3K') 
                              or explicit dimensions (e.g., '2048x2048'). Defaults to None.
        watermark (bool, optional): Whether to add an AI watermark. Defaults to False.
        optimize_prompt_options (dict, optional): Options to optimize the prompt (e.g., {"mode": "standard"}). Defaults to None.
        tools (list, optional): A list of tools for the model to use (e.g., [{"type": "web_search"}]).
        output_format (str, optional): Output image format (e.g., 'png', 'jpeg', 'webp'). Defaults to None.
        sequential_image_generation (str, optional): Sequential image generation strategy (e.g., 'auto'). Defaults to None.

    Environment Variable Reading Logic:
        1. Checks if `api_key` param was explicitly provided to the function.
        2. If missing, attempts to read from the OS environment variable `ARK_API_KEY`.
        3. If still missing, sequentially checks for a `.env` file containing `ARK_API_KEY=` 
           in three locations: `/.env`, `./.env` (current directory), and `<script_directory>/.env`.
        4. If the key is not found after all attempts, returns a dictionary with an error message prompting setup.
        
    Returns:
        dict: The JSON response dictionary.
        
        Successful Return Value Format:
        {
          "data": [
            {
              "url": "https://...",
              "size": "2048x2048" # Included depending on the model capability
            }
          ],
          "usage": { 
              "generated_images": 1,
              ...
          }
        }
        
        Failure Return Value Format:
        {
          "error": "<Error Code or Details>",
          "message": "<Optional detailed message from server>"
        }
    """
    # Retrieve API key
    key_read_from_arg = api_key is not None
    key = api_key or os.environ.get("ARK_API_KEY")
    if not key:
        for path in ["/.env", ".env", sys.path[0] + "/.env"]:
            if os.path.exists(path):
                with open(path, "r") as f:
                    for line in f:
                        if line.startswith("ARK_API_KEY="):
                            key = line.strip().split("=", 1)[1].strip().strip("\"'")
                            break
            if key:
                break
    
    model_id = model or os.environ.get("SEEDREAM_ENDPOINT_ID")
    if not model_id and not key_read_from_arg:
        for path in ["/.env", ".env", sys.path[0] + "/.env"]:
            if os.path.exists(path):
                with open(path, "r") as f:
                    for line in f:
                        if line.startswith("SEEDREAM_ENDPOINT_ID="):
                            model_id = line.strip().split("=", 1)[1].strip().strip("\"'")
                            break
            if model_id:
                break

    if not model_id:
        return {"error": "Require model or SEEDREAM_ENDPOINT_ID in .env/environment"}

    url = "https://ark.cn-beijing.volces.com/api/v3/images/generations"
    payload = {
        "model": model_id,
        "prompt": prompt,
        "watermark": watermark if watermark is not None else False
    }
    
    if size is not None:
        payload["size"] = size
    if optimize_prompt_options is not None:
        payload["optimize_prompt_options"] = optimize_prompt_options
    if tools is not None:
        payload["tools"] = tools
    if output_format is not None:
        payload["output_format"] = output_format
    if sequential_image_generation is not None:
        payload["sequential_image_generation"] = sequential_image_generation

    headers = {
        "Authorization": f"Bearer {key}",
        "Content-Type": "application/json"
    }
    
    req = urllib.request.Request(url, data=json.dumps(payload).encode("utf-8"), headers=headers, method="POST")
    try:
        with urllib.request.urlopen(req) as response:
            res_json = json.loads(response.read().decode("utf-8"))
            
            # If successful, handle downloads if download_dir is provided
            if "data" in res_json:
                res_json["input_params"] = payload
                
                if download_dir:
                    save_path = download_dir
                    if not os.path.exists(save_path):
                        os.makedirs(save_path)
                    
                    for item in res_json["data"]:
                        if "url" in item:
                            try:
                                file_ext = ".png" if payload.get("output_format") == "png" else ".jpeg"
                                filename = f"image_{int(res_json.get('created', 0))}_{res_json['data'].index(item)}{file_ext}"
                                local_file_path = os.path.join(save_path, filename)
                                
                                urllib.request.urlretrieve(item["url"], local_file_path)
                                item["local_path"] = local_file_path
                            except Exception as download_err:
                                item["local_path_error"] = str(download_err)
            
            return res_json
    except urllib.error.HTTPError as e:
        return {"error": e.code, "message": e.read().decode("utf-8")}
    except Exception as e:
        return {"error": str(e)}

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Seedream Image Generation API Wrapper")
    parser.add_argument("--prompt", type=str, required=True, help="Image generation prompt")
    parser.add_argument("--model", type=str, help="Model endpoint id (if not provided, reads SEEDREAM_ENDPOINT_ID)")
    parser.add_argument("--api_key", type=str, help="Volcengine Ark API Key")
    parser.add_argument("--size", type=str, help="Image size")
    parser.add_argument("--watermark", type=str, help="Add watermark (true/false)")
    parser.add_argument("--optimize_prompt_options", type=str, help="JSON string for prompt optimization options")
    parser.add_argument("--tools", type=str, help="JSON string for tools list, e.g. '[{\"type\": \"web_search\"}]'")
    parser.add_argument("--output_format", type=str, help="Output image format (png, jpeg)")
    parser.add_argument("--sequential_image_generation", type=str, help="Sequential image generation mode (e.g., 'auto')")
    parser.add_argument("--download_dir", type=str, help="Directory to save images")
    args = parser.parse_args()
    
    watermark_val = None
    if args.watermark is not None:
        watermark_val = args.watermark.lower() == "true"
        
    optimize_prompt_val = None
    if args.optimize_prompt_options is not None:
        try:
            optimize_prompt_val = json.loads(args.optimize_prompt_options)
        except json.JSONDecodeError:
            print("Error: --optimize_prompt_options must be a valid JSON string.")
            sys.exit(1)

    tools_val = None
    if args.tools is not None:
        try:
            tools_val = json.loads(args.tools)
        except json.JSONDecodeError:
            print("Error: --tools must be a valid JSON string.")
            sys.exit(1)

    res = generate_image(
        prompt=args.prompt,
        model=args.model,
        api_key=args.api_key,
        size=args.size,
        watermark=watermark_val,
        optimize_prompt_options=optimize_prompt_val,
        tools=tools_val,
        output_format=args.output_format,
        sequential_image_generation=args.sequential_image_generation,
        download_dir=args.download_dir
    )
    print(json.dumps(res, indent=2))

File: skills/seedream-image-generation/seedream.js
const https = require('https');
const fs = require('fs');
const path = require('path');

/**
 * Retrieves the API key based on the environment fallback logic.
 * 
 * Environment Variable Reading Logic:
 * 1. Checks the explicitly provided customKey argument.
 * 2. If missing, checks the `ARK_API_KEY` system environment variable.
 * 3. If missing, searches for a `.env` file in the following paths: `/.env`, `./.env` (Current Working Directory), and `<script_directory>/.env`.
 *    It parses the file for the `ARK_API_KEY=` pattern and extracts the value if found.
 * 4. Returns null if all attempts fail.
 * 
 * @param {string|null} customKey - The user-provided API key, if any.
 * @returns {string|null} The resolved API key or null.
 */
/**
 * Retrieves a value from environment or .env files.
 */
function getFromEnv(keyName) {
    if (process.env[keyName]) return process.env[keyName];

    const envPaths = ['/.env', '.env', path.join(process.cwd(), '.env'), path.join(__dirname, '.env')];
    for (const envPath of envPaths) {
        if (fs.existsSync(envPath)) {
            const content = fs.readFileSync(envPath, 'utf8');
            const lines = content.split('\n');
            for (const line of lines) {
                if (line.startsWith(`${keyName}=`)) {
                    let value = line.substring(keyName.length + 1).trim();
                    if (value.startsWith('"') || value.startsWith("'")) {
                        value = value.substring(1, value.length - 1);
                    }
                    return value;
                }
            }
        }
    }
    return null;
}

function getApiKey(customKey) {
    return customKey || getFromEnv('ARK_API_KEY');
}

function getModelId(customModel) {
    return customModel || getFromEnv('SEEDREAM_ENDPOINT_ID');
}

/**
 * Generates an image using the Volcengine Seedream API.
 * 
 * @param {Object} params - The inputs for the generation request.
 * @param {string} params.prompt - The text description of the image to generate. Required.
 * @param {string} [params.model] - The endpoint ID of the model. If missing, uses SEEDREAM_ENDPOINT_ID.
 * @param {string} [params.api_key] - Custom API key to use. Overrides environment variable.
 * @param {string} [params.size] - Target image size. Can be a resolution class (e.g., '2K', '3K') or explicit dimensions (e.g., '2048x2048').
 * @param {boolean} [params.watermark=false] - Whether to include an AI watermark.
 * @param {Object} [params.optimize_prompt_options] - Options to auto-optimize the prompt.
 * @param {Array} [params.tools] - A list of tools for the model to use (e.g., [{"type": "web_search"}]).
 * @param {string} [params.output_format] - Output image format (e.g., 'png', 'jpeg').
 * @param {string} [params.sequential_image_generation] - Sequential image generation strategy (e.g., 'auto').
 * @param {string} [params.download_dir] - Directory to save generated images.
 * 
 * @returns {Promise<Object>} A promise resolving to the API response object.
 * 
 * Return value format (on success):
 * {
 *   "data": [
 *     {
 *       "url": "https://...",
 *       "size": "2048x2048"
 *     }
 *   ],
 *   "input_params": { ... },
 *   "usage": { 
 *       "generated_images": 1,
 *       ...
 *   }
 * }
 * 
 * Return value format (on failure):
 * {
 *   "error": "<Error Details or HTTP Code>", 
 *   "message": "<Optional Server Message>" 
 * }
 */
async function generateImage(params) {
    const modelId = getModelId(params.model);
    if (!modelId) {
        return { error: "Require model or SEEDREAM_ENDPOINT_ID in .env/environment" };
    }

    const payload = {
        model: modelId,
        prompt: params.prompt,
        watermark: params.watermark !== undefined && params.watermark !== null ? params.watermark : false
    };

    if (params.size !== undefined && params.size !== null) {
        payload.size = params.size;
    }
    if (params.optimize_prompt_options !== undefined && params.optimize_prompt_options !== null) {
        payload.optimize_prompt_options = params.optimize_prompt_options;
    }
    if (params.tools !== undefined && params.tools !== null) {
        payload.tools = params.tools;
    }
    if (params.output_format !== undefined && params.output_format !== null) {
        payload.output_format = params.output_format;
    }
    if (params.sequential_image_generation !== undefined && params.sequential_image_generation !== null) {
        payload.sequential_image_generation = params.sequential_image_generation;
    }

    const downloadImage = (url, dest) => {
        return new Promise((resolve, reject) => {
            const file = fs.createWriteStream(dest);
            https.get(url, (response) => {
                response.pipe(file);
                file.on('finish', () => {
                    file.close(resolve);
                });
            }).on('error', (err) => {
                fs.unlink(dest, () => { });
                reject(err.message);
            });
        });
    };

    const dataString = JSON.stringify(payload);

    const options = {
        hostname: 'ark.cn-beijing.volces.com',
        port: 443,
        path: '/api/v3/images/generations',
        method: 'POST',
        headers: {
            'Authorization': `Bearer ${apiKey}`,
            'Content-Type': 'application/json',
            'Content-Length': Buffer.byteLength(dataString)
        }
    };

    return new Promise((resolve) => {
        const req = https.request(options, (res) => {
            let resData = '';
            res.on('data', (chunk) => resData += chunk);
            res.on('end', async () => {
                try {
                    const resJson = JSON.parse(resData);

                    if (resJson.data) {
                        resJson.input_params = payload;

                        if (params.download_dir) {
                            const saveDir = params.download_dir;
                            if (!fs.existsSync(saveDir)) {
                                fs.mkdirSync(saveDir, { recursive: true });
                            }

                            for (let i = 0; i < resJson.data.length; i++) {
                                const item = resJson.data[i];
                                if (item.url) {
                                    try {
                                        const ext = payload.output_format === 'png' ? '.png' : '.jpeg';
                                        const filename = `image_${resJson.created || Date.now()}_${i}${ext}`;
                                        const destPath = path.join(saveDir, filename);
                                        await downloadImage(item.url, destPath);
                                        item.local_path = destPath;
                                    } catch (err) {
                                        item.local_path_error = err;
                                    }
                                }
                            }
                        }
                    }
                    resolve(resJson);
                } catch (e) {
                    resolve({ error: "Parse error", message: resData });
                }
            });
        });

        req.on('error', (e) => resolve({ error: e.message }));
        req.write(dataString);
        req.end();
    });
}

if (require.main === module) {
    const args = process.argv.slice(2);
    const params = {};
    for (let i = 0; i < args.length; i++) {
        if (args[i].startsWith('--')) {
            const key = args[i].substring(2);
            params[key] = args[i + 1];
            i++;
        }
    }

    if (params.watermark === "true") params.watermark = true;
    if (params.watermark === "false") params.watermark = false;

    if (params.optimize_prompt_options) {
        try {
            params.optimize_prompt_options = JSON.parse(params.optimize_prompt_options);
        } catch (e) {
            console.error("Error: --optimize_prompt_options must be a valid JSON string");
            process.exit(1);
        }
    }

    if (params.tools) {
        try {
            params.tools = JSON.parse(params.tools);
        } catch (e) {
            console.error("Error: --tools must be a valid JSON string");
            process.exit(1);
        }
    }

    if (!params.prompt) {
        console.error("Usage: node seedream.js --prompt <prompt> [--model <model>] [--api_key <key>] [--size <size>] [--watermark <true/false>] [--output_format <format>] [--sequential_image_generation <mode>] [--tools <json>] [--optimize_prompt_options <json>]");
        process.exit(1);
    }

    generateImage(params).then(res => console.log(JSON.stringify(res, null, 2)));
}

module.exports = { generateImage };
